### PR TITLE
PcaTrainer as estimator

### PR DESCRIPTION
**Pull request by sfilipi**
Addresses part of #754

### CODE REVIEW

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 1]
```diff
@@ -108,6 +137,18 @@ public override PcaPredictor Train(TrainContext context)
             }
         }
 
+        private static SchemaShape.Column MakeWeightColumn(string weightColumn)
```
Comment (sfilipi) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: Use trainerUtils

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 2]
```diff
@@ -266,6 +307,27 @@ private static void PostProcess(VBuffer<Float>[] y, Float[] sigma, Float[] z, in
...
+                new SchemaShape.Column(DefaultColumnNames.PredictedLabel,
+                        SchemaShape.Column.VectorKind.Scalar,
+                        BoolType.Instance,
+                        false,
+                        new SchemaShape(MetadataUtils.GetTrainerOutputMetadata()))
+            };
+        }
+
+        protected override BinaryPredictionTransformer<PcaPredictor> MakeTransformer(PcaPredictor model, ISchema trainSchema)
+            => new BinaryPredictionTransformer<PcaPredictor>(Host, model, trainSchema, _featureColumn);
```
Comment (sfilipi) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: >a, _featu [](start = 84, length = 9)

@Zruty do we have a threeshold, to pass?

**Comment by sfilipi** [comment]
>             Model = model;

compact all constructors here. 

---
Refers to: src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs:71 in c1d54f7. [](commit_id = c1d54f79de499475a23bfc155eefcc879dddd581, deletion_comment = False)

File: src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs [hunk 3]
```diff
@@ -91,9 +95,23 @@ protected PredictionTransformerBase(IHost host, ModelLoadContext ctx)
         /// </summary>
         /// <param name="input">The input data.</param>
         /// <returns>The transformed <see cref="IDataView"/></returns>
-        public abstract IDataView Transform(IDataView input);
 
-        public abstract IRowToRowMapper GetRowToRowMapper(ISchema inputSchema);
+        public IDataView Transform(IDataView input)
+        {
+            Host.CheckValue(input, nameof(input));
+            return Scorer.ApplyToData(Host, input);
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs]: >Scorer [](start = 19, length = 6)

SDCA, FastTree, LightGbm, GAM don't have Scorer setup, so tests are failing.

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 2 above)
Comment (Zruty0) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: we use default

---
In reply to: [219663367](https://github.com/dotnet/machinelearning/pull/996#discussion_r219663367) [](ancestors = 219663367)

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 4]
```diff
@@ -41,7 +42,7 @@ namespace Microsoft.ML.Runtime.PCA
     /// <remarks>
     /// This PCA can be made into Kernel PCA by using Random Fourier Features transform
     /// </remarks>
-    public sealed class RandomizedPcaTrainer : TrainerBase<PcaPredictor>
+    public sealed class RandomizedPcaTrainer : TrainerEstimatorBase<BinaryPredictionTransformer<PcaPredictor>, PcaPredictor>
```
Comment (TomFinley) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: >BinaryPredictionTransformer [](start = 68, length = 27)

So, why is this a binary prediction transformer? Prior to your change it was not a binary predictor, it was part of the anomaly detection task.

File: src/Microsoft.ML.Data/Scorers/BinaryClassifierScorer.cs [hunk 5]
```diff
@@ -297,8 +297,6 @@ private static ColumnType GetPredColType(ColumnType scoreType, ISchemaBoundRowMa
         }
 
         private static bool OutputTypeMatches(ColumnType scoreType)
-        {
-            return scoreType == NumberType.Float;
-        }
+        => scoreType == NumberType.Float;
     }
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/Scorers/BinaryClassifierScorer.cs]: Bit of a nit, but while you're at it, maybe indent properly.

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 4 above)
Comment (sfilipi) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: Good point, thanks. 

---
In reply to: [220048006](https://github.com/dotnet/machinelearning/pull/996#discussion_r220048006) [](ancestors = 220048006)

**Comment by sfilipi** [comment]
Closing and re-opening to trigger rebuild. The local build is fine..

File: src/Microsoft.ML.PCA/PcaTrainer.cs [hunk 6]
```diff
@@ -73,28 +74,66 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
...
+        /// <summary>
+        /// Initializes a new instance of <see cref="RandomizedPcaTrainer"/>.
+        /// </summary>
+        /// <param name="env">The local instance of the <see cref="IHostEnvironment"/>.</param>
+        /// <param name="featureColumn">The name of the feature column.</param>
+        /// <param name="weightColumn">The name of the weight column.</param>
+        /// <param name="rank">The number of components in the PCA.</param>
+        /// <param name="oversampling">Oversampling parameter for randomized PCA training.</param>
+        /// <param name="center">If enabled, data is centered to be zero mean.</param>
+        /// <param name="seed">The seed for random number generation.</param>
```
Comment (Zruty0) [inline src/Microsoft.ML.PCA/PcaTrainer.cs]: >seed [](start = 25, length = 4)

why do we use a separate seed here?

File: test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs [hunk 7]
```diff
@@ -17,6 +18,32 @@ public TrainerEstimators(ITestOutputHelper helper) : base(helper)
...
+                }
+            });
+            var data = reader.Read(new MultiFileSource(GetDataPath(TestDatasets.mnistOneClass.trainFilename)));
+
+
+            // Pipeline.
+            var pipeline = new RandomizedPcaTrainer(Env, featureColumn, rank:10);
+
+            TestEstimatorCore(pipeline, data);
+        }
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs]: > } [](start = 7, length = 2)

`Done()`

File: src/Microsoft.ML.PCA/PcaTrainer.cs (hunk 6 above)
Comment (sfilipi) [reply src/Microsoft.ML.PCA/PcaTrainer.cs]: it is part of the args, therefore i moved it here; rather than doing the delegate on advancedArgs. 

---
In reply to: [220353522](https://github.com/dotnet/machinelearning/pull/996#discussion_r220353522) [](ancestors = 220353522)
